File: backend/app/models/tax.py
from decimal import Decimal
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, ForeignKey, Numeric, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base
from app.models.user import TimestampMixin

if TYPE_CHECKING:
    from app.models.restaurant import Restaurant


class RestaurantTaxSettings(Base, TimestampMixin):
    __tablename__ = "restaurant_tax_settings"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    restaurant_id: Mapped[int] = mapped_column(
        ForeignKey("restaurants.id"), unique=True, nullable=False
    )
    tax_enabled: Mapped[bool] = mapped_column(default=True, nullable=False)
    tax_rate: Mapped[Decimal] = mapped_column(Numeric(5, 4), default=Decimal("0.08"), nullable=False)
    tax_label: Mapped[str] = mapped_column(String(50), default="Sales Tax", nullable=False)
    service_charge_enabled: Mapped[bool] = mapped_column(default=False, nullable=False)
    service_charge_rate: Mapped[Decimal] = mapped_column(
        Numeric(5, 4), default=Decimal("0.00"), nullable=False
    )
    service_charge_label: Mapped[str] = mapped_column(
        String(50), default="Service Charge", nullable=False
    )
    prices_include_tax: Mapped[bool] = mapped_column(default=False, nullable=False)

    restaurant: Mapped["Restaurant"] = relationship(back_populates="tax_settings")

File: src/types/tax.ts
export interface TaxSettings {
  id: number;
  restaurant_id: number;
  tax_enabled: boolean;
  tax_rate: number;
  tax_label: string;
  service_charge_enabled: boolean;
  service_charge_rate: number;
  service_charge_label: string;
  prices_include_tax: boolean;
}

export interface TaxSettingsUpdate {
  tax_enabled?: boolean;
  tax_rate?: number;
  tax_label?: string;
  service_charge_enabled?: boolean;
  service_charge_rate?: number;
  service_charge_label?: string;
  prices_include_tax?: boolean;
}

export interface OrderTotals {
  subtotal: number;
  tax: number;
  serviceCharge: number;
  total: number;
  taxLabel: string;
  serviceChargeLabel: string;
  showTax: boolean;
  showServiceCharge: boolean;
}
